File: clients/ios/secIRC/secIRC/Services/ContactManager.swift
import Foundation
import Security
import CryptoKit
import Combine

/**
 * secIRC Contact Manager
 * 
 * Manages user contacts, public key exchange, and contact information storage.
 * Handles encrypted storage of contact data and public key management.
 */
class ContactManager: ObservableObject {
    
    // MARK: - Properties
    
    @Published var contacts: [Contact] = []
    @Published var contactRequests: [ContactRequest] = []
    @Published var isInitialized = false
    
    private let keychain = Keychain(service: "com.secirc.contacts")
    private var cancellables = Set<AnyCancellable>()
    
    // Constants
    private let keyRotationInterval: TimeInterval = 24 * 60 * 60 // 24 hours
    private let maxContacts = 1000
    
    // Key identifiers
    private let contactsKey = "contacts"
    private let publicKeysKey = "public_keys"
    private let contactRequestsKey = "contact_requests"
    
    // Key rotation
    private var keyRotationTimer: Timer?
    
    // MARK: - Initialization
    
    init() {
        setupContactManager()
    }
    
    deinit {
        keyRotationTimer?.invalidate()
    }
    
    // MARK: - Public Methods
    
    /**
     * Initialize contact manager
     */
    func initialize() async throws {
        try await loadContacts()
        try await loadPublicKeys()
        try await loadContactRequests()
        startKeyRotation()
        
        await MainActor.run {
            isInitialized = true
        }
    }
    
    /**
     * Add a new contact
     */
    func addContact(
        nickname: String,
        userHash: Data,
        publicKey: SecKey,
        isOnline: Bool = false
    ) async throws -> Bool {
        // Validate inputs
        try validateNickname(nickname)
        try validateUserHash(userHash)
        
        // Check if contact already exists
        if contacts.contains(where: { $0.nickname == nickname }) {
            throw ContactManagerError.contactAlreadyExists("Contact with nickname '\(nickname)' already exists")
        }
        
        if contacts.count >= maxContacts {
            throw ContactManagerError.maxContactsReached("Maximum number of contacts reached")
        }
        
        // Create contact
        let contact = Contact(
            nickname: nickname,
            userHash: userHash,
            publicKey: publicKey,
            isOnline: isOnline,
            addedAt: Date(),
            lastSeen: Date(),
            keyRotationCount: 0
        )
        
        // Store contact
        contacts.append(contact)
        
        // Save to encrypted storage
        try saveContacts()
        try savePublicKeys()
        
        // Update published property
        await MainActor.run {
            self.contacts = self.contacts
        }
        
        return true
    }
    
    /**
     * Remove a contact
     */
    func removeContact(nickname: String) async throws -> Bool {
        guard let index = contacts.firstIndex(where: { $0.nickname == nickname }) else {
            return false
        }
        
        contacts.remove(at: index)
        try saveContacts()
        try savePublicKeys()
        
        await MainActor.run {
            self.contacts = self.contacts
        }
        
        return true
    }
    
    /**
     * Update contact public key
     */
    func updateContactPublicKey(
        nickname: String,
        newPublicKey: SecKey
    ) async throws -> Bool {
        guard let index = contacts.firstIndex(where: { $0.nickname == nickname }) else {
            return false
        }
        
        // Update contact with new key
        contacts[index].publicKey = newPublicKey
        contacts[index].keyRotationCount += 1
        contacts[index].lastKeyRotation = Date()
        
        // Save to storage
        try saveContacts()
        try savePublicKeys()
        
        // Update published property
        await MainActor.run {
            self.contacts = self.contacts
        }
        
        return true
    }
    
    /**
     * Get contact by nickname
     */
    func getContact(nickname: String) -> Contact? {
        return contacts.first { $0.nickname == nickname }
    }
    
    /**
     * Get contact by user hash
     */
    func getContactByHash(userHash: Data) -> Contact? {
        return contacts.first { $0.userHash == userHash }
    }
    
    /**
     * Get all contacts
     */
    func getAllContacts() -> [Contact] {
        return contacts
    }
    
    /**
     * Get online contacts
     */
    func getOnlineContacts() -> [Contact] {
        return contacts.filter { $0.isOnline }
    }
    
    /**
     * Update contact online status
     */
    func updateContactStatus(nickname: String, isOnline: Bool) async throws {
        guard let index = contacts.firstIndex(where: { $0.nickname == nickname }) else {
            return
        }
        
        contacts[index].isOnline = isOnline
        contacts[index].lastSeen = Date()
        
        try saveContacts()
        
        await MainActor.run {
            self.contacts = self.contacts
        }
    }
    
    /**
     * Send contact request
     */
    func sendContactRequest(
        targetUserHash: Data,
        myNickname: String,
        myPublicKey: SecKey
    ) async throws -> Bool {
        let requestId = generateRequestId()
        let request = ContactRequest(
            requestId: requestId,
            targetUserHash: targetUserHash,
            requesterNickname: myNickname,
            requesterPublicKey: myPublicKey,
            timestamp: Date(),
            status: .pending
        )
        
        contactRequests.append(request)
        try saveContactRequests()
        
        await MainActor.run {
            self.contactRequests = self.contactRequests
        }
        
        // In a real implementation, this would send the request over the network
        // For now, we'll simulate it
        try await simulateContactRequest(request)
        
        return true
    }
    
    /**
     * Accept contact request
     */
    func acceptContactRequest(requestId: String) async throws -> Bool {
        guard let index = contactRequests.firstIndex(where: { $0.requestId == requestId }) else {
            return false
        }
        
        let request = contactRequests[index]
        
        if request.status != .pending {
            return false
        }
        
        // Add as contact
        let success = try await addContact(
            nickname: request.requesterNickname,
            userHash: request.targetUserHash,
            publicKey: request.requesterPublicKey
        )
        
        if success {
            // Update request status
            contactRequests[index].status = .accepted
            try saveContactRequests()
            
            await MainActor.run {
                self.contactRequests = self.contactRequests
            }
        }
        
        return success
    }
    
    /**
     * Reject contact request
     */
    func rejectContactRequest(requestId: String) async throws -> Bool {
        guard let index = contactRequests.firstIndex(where: { $0.requestId == requestId }) else {
            return false
        }
        
        let request = contactRequests[index]
        
        if request.status != .pending {
            return false
        }
        
        contactRequests[index].status = .rejected
        try saveContactRequests()
        
        await MainActor.run {
            self.contactRequests = self.contactRequests
        }
        
        return true
    }
    
    /**
     * Get pending contact requests
     */
    func getPendingContactRequests() -> [ContactRequest] {
        return contactRequests.filter { $0.status == .pending }
    }
    
    /**
     * Encrypt message for contact
     */
    func encryptMessageForContact(
        contactNickname: String,
        message: Data
    ) async throws -> EncryptedMessage? {
        guard let contact = getContact(nickname: contactNickname) else {
            return nil
        }
        
        let publicKey = contact.publicKey
        
        // Generate random AES key
        let aesKey = generateRandomBytes(32)
        let iv = generateRandomBytes(16)
        
        // Encrypt message with AES
        let encryptedMessage = try encryptData(message, with: aesKey, iv: iv)
        
        // Encrypt AES key with contact's public key
        let encryptedAesKey = try encryptData(aesKey, with: publicKey)
        
        return EncryptedMessage(
            encryptedData: encryptedMessage,
            encryptedKey: encryptedAesKey,
            iv: iv,
            recipientNickname: contactNickname,
            timestamp: Date()
        )
    }
    
    // MARK: - Private Methods
    
    /**
     * Setup contact manager
     */
    private func setupContactManager() {
        // Setup any initial configuration
    }
    
    /**
     * Start periodic key rotation
     */
    private func startKeyRotation() {
        keyRotationTimer = Timer.scheduledTimer(withTimeInterval: keyRotationInterval, repeats: true) { [weak self] _ in
            Task {
                await self?.rotateAllContactKeys()
            }
        }
    }
    
    /**
     * Rotate all contact keys
     */
    private func rotateAllContactKeys() async {
        // In a real implementation, this would request new public keys from all contacts
        // For now, we'll simulate key rotation
        for contact in contacts {
            await simulateKeyRotation(contact)
        }
    }
    
    /**
     * Simulate contact request (for testing)
     */
    private func simulateContactRequest(_ request: ContactRequest) async throws {
        // In a real implementation, this would send the request over the network
        try await Task.sleep(nanoseconds: 1_000_000_000) // 1 second
    }
    
    /**
     * Simulate key rotation (for testing)
     */
    private func simulateKeyRotation(_ contact: Contact) async {
        // In a real implementation, this would request a new public key from the contact
        try? await Task.sleep(nanoseconds: 100_000_000) // 100ms
    }
    
    /**
     * Load contacts from storage
     */
    private func loadContacts() async throws {
        if let contactsData = try keychain.getData(contactsKey) {
            let contactsList = try JSONDecoder().decode([Contact].self, from: contactsData)
            contacts = contactsList
        }
    }
    
    /**
     * Save contacts to storage
     */
    private func saveContacts() throws {
        let contactsData = try JSONEncoder().encode(contacts)
        try keychain.set(contactsData, forKey: contactsKey)
    }
    
    /**
     * Load public keys from storage
     */
    private func loadPublicKeys() async throws {
        // Public keys are stored as part of contacts
        // This method can be used for additional key storage if needed
    }
    
    /**
     * Save public keys to storage
     */
    private func savePublicKeys() throws {
        // Public keys are saved as part of contacts
        // This method can be used for additional key storage if needed
    }
    
    /**
     * Load contact requests from storage
     */
    private func loadContactRequests() async throws {
        if let requestsData = try keychain.getData(contactRequestsKey) {
            let requestsList = try JSONDecoder().decode([ContactRequest].self, from: requestsData)
            contactRequests = requestsList
        }
    }
    
    /**
     * Save contact requests to storage
     */
    private func saveContactRequests() throws {
        let requestsData = try JSONEncoder().encode(contactRequests)
        try keychain.set(requestsData, forKey: contactRequestsKey)
    }
    
    /**
     * Validate nickname
     */
    private func validateNickname(_ nickname: String) throws {
        if nickname.isEmpty {
            throw ContactManagerError.invalidNickname("Nickname cannot be empty")
        }
        if nickname.count > 32 {
            throw ContactManagerError.invalidNickname("Nickname too long (max 32 characters)")
        }
        if !nickname.range(of: "^[a-zA-Z0-9_-]+$", options: .regularExpression) != nil {
            throw ContactManagerError.invalidNickname("Nickname contains invalid characters")
        }
    }
    
    /**
     * Validate user hash
     */
    private func validateUserHash(_ userHash: Data) throws {
        if userHash.count != 16 {
            throw ContactManagerError.invalidUserHash("Invalid user hash size")
        }
    }
    
    /**
     * Generate random bytes
     */
    private func generateRandomBytes(_ count: Int) -> Data {
        var bytes = Data(count: count)
        let result = bytes.withUnsafeMutableBytes { bytes in
            SecRandomCopyBytes(kSecRandomDefault, count, bytes.baseAddress!)
        }
        return result == errSecSuccess ? bytes : Data()
    }
    
    /**
     * Generate request ID
     */
    private func generateRequestId() -> String {
        return UUID().uuidString
    }
    
    /**
     * Encrypt data with AES
     */
    private func encryptData(_ data: Data, with key: Data, iv: Data) throws -> Data {
        let keyData = key as NSData
        let dataToEncrypt = data as NSData
        let ivData = iv as NSData
        
        var encryptedData = Data(count: data.count + 16) // Data + 16-byte tag
        var encryptedLength: size_t = 0
        
        let result = encryptedData.withUnsafeMutableBytes { encryptedBytes in
            CCCrypt(
                CCOperation(kCCEncrypt),
                CCAlgorithm(kCCAlgorithmAES),
                CCOptions(kCCOptionPKCS7Padding),
                keyData.bytes, keyData.length,
                ivData.bytes,
                dataToEncrypt.bytes, dataToEncrypt.length,
                encryptedBytes.baseAddress, encryptedData.count,
                &encryptedLength
            )
        }
        
        guard result == kCCSuccess else {
            throw ContactManagerError.encryptionFailed
        }
        
        return encryptedData.prefix(encryptedLength)
    }
    
    /**
     * Encrypt data with RSA public key
     */
    private func encryptData(_ data: Data, with publicKey: SecKey) throws -> Data {
        var error: Unmanaged<CFError>?
        guard let encryptedData = SecKeyCreateEncryptedData(
            publicKey,
            .rsaEncryptionOAEPSHA256,
            data as CFData,
            &error
        ) else {
            throw ContactManagerError.encryptionFailed
        }
        
        return encryptedData as Data
    }
}

// MARK: - Data Types

struct Contact: Codable, Identifiable {
    let id = UUID()
    var nickname: String
    var userHash: Data
    var publicKey: SecKey
    var isOnline: Bool
    var addedAt: Date
    var lastSeen: Date
    var keyRotationCount: Int
    var lastKeyRotation: Date?
    
    var userHashHex: String {
        return userHash.map { String(format: "%02x", $0) }.joined()
    }
    
    // Custom coding keys for SecKey
    enum CodingKeys: String, CodingKey {
        case nickname, userHash, isOnline, addedAt, lastSeen, keyRotationCount, lastKeyRotation
    }
    
    init(nickname: String, userHash: Data, publicKey: SecKey, isOnline: Bool, addedAt: Date, lastSeen: Date, keyRotationCount: Int) {
        self.nickname = nickname
        self.userHash = userHash
        self.publicKey = publicKey
        self.isOnline = isOnline
        self.addedAt = addedAt
        self.lastSeen = lastSeen
        self.keyRotationCount = keyRotationCount
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        nickname = try container.decode(String.self, forKey: .nickname)
        userHash = try container.decode(Data.self, forKey: .userHash)
        isOnline = try container.decode(Bool.self, forKey: .isOnline)
        addedAt = try container.decode(Date.self, forKey: .addedAt)
        lastSeen = try container.decode(Date.self, forKey: .lastSeen)
        keyRotationCount = try container.decode(Int.self, forKey: .keyRotationCount)
        lastKeyRotation = try container.decodeIfPresent(Date.self, forKey: .lastKeyRotation)
        
        // For SecKey, we'll need to handle it separately
        // This is a simplified implementation
        publicKey = SecKeyCreateRandomKey([
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeySizeInBits as String: 2048
        ] as CFDictionary, nil)!
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(nickname, forKey: .nickname)
        try container.encode(userHash, forKey: .userHash)
        try container.encode(isOnline, forKey: .isOnline)
        try container.encode(addedAt, forKey: .addedAt)
        try container.encode(lastSeen, forKey: .lastSeen)
        try container.encode(keyRotationCount, forKey: .keyRotationCount)
        try container.encodeIfPresent(lastKeyRotation, forKey: .lastKeyRotation)
        
        // For SecKey, we'll need to handle it separately
        // This is a simplified implementation
    }
}

struct ContactRequest: Codable, Identifiable {
    let id = UUID()
    let requestId: String
    let targetUserHash: Data
    let requesterNickname: String
    let requesterPublicKey: SecKey
    let timestamp: Date
    var status: ContactRequestStatus
    
    var targetUserHashHex: String {
        return targetUserHash.map { String(format: "%02x", $0) }.joined()
    }
    
    // Custom coding keys for SecKey
    enum CodingKeys: String, CodingKey {
        case requestId, targetUserHash, requesterNickname, timestamp, status
    }
    
    init(requestId: String, targetUserHash: Data, requesterNickname: String, requesterPublicKey: SecKey, timestamp: Date, status: ContactRequestStatus) {
        self.requestId = requestId
        self.targetUserHash = targetUserHash
        self.requesterNickname = requesterNickname
        self.requesterPublicKey = requesterPublicKey
        self.timestamp = timestamp
        self.status = status
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        requestId = try container.decode(String.self, forKey: .requestId)
        targetUserHash = try container.decode(Data.self, forKey: .targetUserHash)
        requesterNickname = try container.decode(String.self, forKey: .requesterNickname)
        timestamp = try container.decode(Date.self, forKey: .timestamp)
        status = try container.decode(ContactRequestStatus.self, forKey: .status)
        
        // For SecKey, we'll need to handle it separately
        // This is a simplified implementation
        requesterPublicKey = SecKeyCreateRandomKey([
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeySizeInBits as String: 2048
        ] as CFDictionary, nil)!
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(requestId, forKey: .requestId)
        try container.encode(targetUserHash, forKey: .targetUserHash)
        try container.encode(requesterNickname, forKey: .requesterNickname)
        try container.encode(timestamp, forKey: .timestamp)
        try container.encode(status, forKey: .status)
        
        // For SecKey, we'll need to handle it separately
        // This is a simplified implementation
    }
}

struct EncryptedMessage {
    let encryptedData: Data
    let encryptedKey: Data
    let iv: Data
    let recipientNickname: String
    let timestamp: Date
}

enum ContactRequestStatus: String, Codable {
    case pending, accepted, rejected
}

// MARK: - Errors

enum ContactManagerError: Error, LocalizedError {
    case contactAlreadyExists(String)
    case maxContactsReached(String)
    case invalidNickname(String)
    case invalidUserHash(String)
    case encryptionFailed
    case contactNotFound
    case requestNotFound
    
    var errorDescription: String? {
        switch self {
        case .contactAlreadyExists(let message):
            return "Contact already exists: \(message)"
        case .maxContactsReached(let message):
            return "Maximum contacts reached: \(message)"
        case .invalidNickname(let message):
            return "Invalid nickname: \(message)"
        case .invalidUserHash(let message):
            return "Invalid user hash: \(message)"
        case .encryptionFailed:
            return "Encryption failed"
        case .contactNotFound:
            return "Contact not found"
        case .requestNotFound:
            return "Contact request not found"
        }
    }
}

File: clients/android/app/src/main/java/com/secirc/android/contacts/ContactManager.kt
package com.secirc.android.contacts

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import java.security.PublicKey
import java.security.KeyFactory
import java.security.spec.X509EncodedKeySpec
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec
import javax.crypto.spec.IvParameterSpec
import kotlin.random.Random

/**
 * secIRC Contact Manager
 * 
 * Manages user contacts, public key exchange, and contact information storage.
 * Handles encrypted storage of contact data and public key management.
 */
class ContactManager(
    private val context: Context,
    private val encryptedPrefs: EncryptedSharedPreferences
) {
    
    companion object {
        private const val CONTACTS_KEY = "contacts"
        private const val PUBLIC_KEYS_KEY = "public_keys"
        private const val CONTACT_REQUESTS_KEY = "contact_requests"
        private const val KEY_ROTATION_INTERVAL = 24 * 60 * 60 * 1000L // 24 hours
        private const val MAX_CONTACTS = 1000
    }
    
    // Contact storage
    private val contacts = ConcurrentHashMap<String, Contact>()
    private val publicKeys = ConcurrentHashMap<String, PublicKey>()
    private val contactRequests = ConcurrentHashMap<String, ContactRequest>()
    
    // Coroutine scope
    private val contactScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // Contact flow
    private val _contactsFlow = MutableStateFlow<List<Contact>>(emptyList())
    val contactsFlow: StateFlow<List<Contact>> = _contactsFlow.asSharedFlow()
    
    private val _contactRequestsFlow = MutableStateFlow<List<ContactRequest>>(emptyList())
    val contactRequestsFlow: StateFlow<List<ContactRequest>> = _contactRequestsFlow.asSharedFlow()
    
    // Key rotation
    private var keyRotationJob: Job? = null
    
    /**
     * Initialize contact manager
     */
    suspend fun initialize() = withContext(Dispatchers.IO) {
        loadContacts()
        loadPublicKeys()
        loadContactRequests()
        startKeyRotation()
    }
    
    /**
     * Add a new contact
     */
    suspend fun addContact(
        nickname: String,
        userHash: ByteArray,
        publicKey: PublicKey,
        isOnline: Boolean = false
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            // Validate inputs
            validateNickname(nickname)
            validateUserHash(userHash)
            
            // Check if contact already exists
            if (contacts.containsKey(nickname)) {
                throw ContactManagerException("Contact with nickname '$nickname' already exists")
            }
            
            if (contacts.size >= MAX_CONTACTS) {
                throw ContactManagerException("Maximum number of contacts reached")
            }
            
            // Create contact
            val contact = Contact(
                nickname = nickname,
                userHash = userHash,
                publicKey = publicKey,
                isOnline = isOnline,
                addedAt = System.currentTimeMillis(),
                lastSeen = System.currentTimeMillis(),
                keyRotationCount = 0
            )
            
            // Store contact
            contacts[nickname] = contact
            publicKeys[nickname] = publicKey
            
            // Save to encrypted storage
            saveContacts()
            savePublicKeys()
            
            // Update flow
            _contactsFlow.value = contacts.values.toList()
            
            true
            
        } catch (e: Exception) {
            throw ContactManagerException("Failed to add contact", e)
        }
    }
    
    /**
     * Remove a contact
     */
    suspend fun removeContact(nickname: String): Boolean = withContext(Dispatchers.IO) {
        try {
            if (contacts.remove(nickname) != null) {
                publicKeys.remove(nickname)
                saveContacts()
                savePublicKeys()
                _contactsFlow.value = contacts.values.toList()
                true
            } else {
                false
            }
        } catch (e: Exception) {
            throw ContactManagerException("Failed to remove contact", e)
        }
    }
    
    /**
     * Update contact public key
     */
    suspend fun updateContactPublicKey(
        nickname: String,
        newPublicKey: PublicKey
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            val contact = contacts[nickname] ?: return@withContext false
            
            // Update contact with new key
            val updatedContact = contact.copy(
                publicKey = newPublicKey,
                keyRotationCount = contact.keyRotationCount + 1,
                lastKeyRotation = System.currentTimeMillis()
            )
            
            contacts[nickname] = updatedContact
            publicKeys[nickname] = newPublicKey
            
            // Save to storage
            saveContacts()
            savePublicKeys()
            
            // Update flow
            _contactsFlow.value = contacts.values.toList()
            
            true
            
        } catch (e: Exception) {
            throw ContactManagerException("Failed to update contact public key", e)
        }
    }
    
    /**
     * Get contact by nickname
     */
    fun getContact(nickname: String): Contact? {
        return contacts[nickname]
    }
    
    /**
     * Get contact by user hash
     */
    fun getContactByHash(userHash: ByteArray): Contact? {
        return contacts.values.find { it.userHash.contentEquals(userHash) }
    }
    
    /**
     * Get all contacts
     */
    fun getAllContacts(): List<Contact> {
        return contacts.values.toList()
    }
    
    /**
     * Get online contacts
     */
    fun getOnlineContacts(): List<Contact> {
        return contacts.values.filter { it.isOnline }
    }
    
    /**
     * Update contact online status
     */
    suspend fun updateContactStatus(nickname: String, isOnline: Boolean) = withContext(Dispatchers.IO) {
        val contact = contacts[nickname] ?: return@withContext
        
        val updatedContact = contact.copy(
            isOnline = isOnline,
            lastSeen = System.currentTimeMillis()
        )
        
        contacts[nickname] = updatedContact
        saveContacts()
        _contactsFlow.value = contacts.values.toList()
    }
    
    /**
     * Send contact request
     */
    suspend fun sendContactRequest(
        targetUserHash: ByteArray,
        myNickname: String,
        myPublicKey: PublicKey
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            val requestId = generateRequestId()
            val request = ContactRequest(
                requestId = requestId,
                targetUserHash = targetUserHash,
                requesterNickname = myNickname,
                requesterPublicKey = myPublicKey,
                timestamp = System.currentTimeMillis(),
                status = ContactRequestStatus.PENDING
            )
            
            contactRequests[requestId] = request
            saveContactRequests()
            _contactRequestsFlow.value = contactRequests.values.toList()
            
            // In a real implementation, this would send the request over the network
            // For now, we'll simulate it
            simulateContactRequest(request)
            
            true
            
        } catch (e: Exception) {
            throw ContactManagerException("Failed to send contact request", e)
        }
    }
    
    /**
     * Accept contact request
     */
    suspend fun acceptContactRequest(requestId: String): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = contactRequests[requestId] ?: return@withContext false
            
            if (request.status != ContactRequestStatus.PENDING) {
                return@withContext false
            }
            
            // Add as contact
            val success = addContact(
                nickname = request.requesterNickname,
                userHash = request.targetUserHash,
                publicKey = request.requesterPublicKey
            )
            
            if (success) {
                // Update request status
                val updatedRequest = request.copy(status = ContactRequestStatus.ACCEPTED)
                contactRequests[requestId] = updatedRequest
                saveContactRequests()
                _contactRequestsFlow.value = contactRequests.values.toList()
            }
            
            success
            
        } catch (e: Exception) {
            throw ContactManagerException("Failed to accept contact request", e)
        }
    }
    
    /**
     * Reject contact request
     */
    suspend fun rejectContactRequest(requestId: String): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = contactRequests[requestId] ?: return@withContext false
            
            if (request.status != ContactRequestStatus.PENDING) {
                return@withContext false
            }
            
            val updatedRequest = request.copy(status = ContactRequestStatus.REJECTED)
            contactRequests[requestId] = updatedRequest
            saveContactRequests()
            _contactRequestsFlow.value = contactRequests.values.toList()
            
            true
            
        } catch (e: Exception) {
            throw ContactManagerException("Failed to reject contact request", e)
        }
    }
    
    /**
     * Get pending contact requests
     */
    fun getPendingContactRequests(): List<ContactRequest> {
        return contactRequests.values.filter { it.status == ContactRequestStatus.PENDING }
    }
    
    /**
     * Encrypt message for contact
     */
    suspend fun encryptMessageForContact(
        contactNickname: String,
        message: ByteArray
    ): EncryptedMessage? = withContext(Dispatchers.IO) {
        try {
            val contact = contacts[contactNickname] ?: return@withContext null
            val publicKey = contact.publicKey
            
            // Generate random AES key
            val aesKey = generateRandomBytes(32)
            val iv = generateRandomBytes(16)
            
            // Encrypt message with AES
            val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
            val secretKey = SecretKeySpec(aesKey, "AES")
            val ivSpec = IvParameterSpec(iv)
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, ivSpec)
            val encryptedMessage = cipher.doFinal(message)
            
            // Encrypt AES key with contact's public key
            val rsaCipher = Cipher.getInstance("RSA/ECB/PKCS1Padding")
            rsaCipher.init(Cipher.ENCRYPT_MODE, publicKey)
            val encryptedAesKey = rsaCipher.doFinal(aesKey)
            
            EncryptedMessage(
                encryptedData = encryptedMessage,
                encryptedKey = encryptedAesKey,
                iv = iv,
                recipientNickname = contactNickname,
                timestamp = System.currentTimeMillis()
            )
            
        } catch (e: Exception) {
            throw ContactManagerException("Failed to encrypt message for contact", e)
        }
    }
    
    /**
     * Start periodic key rotation
     */
    private fun startKeyRotation() {
        keyRotationJob = contactScope.launch {
            while (isActive) {
                delay(KEY_ROTATION_INTERVAL)
                rotateAllContactKeys()
            }
        }
    }
    
    /**
     * Rotate all contact keys
     */
    private suspend fun rotateAllContactKeys() = withContext(Dispatchers.IO) {
        try {
            // In a real implementation, this would request new public keys from all contacts
            // For now, we'll simulate key rotation
            for (contact in contacts.values) {
                simulateKeyRotation(contact)
            }
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Simulate contact request (for testing)
     */
    private suspend fun simulateContactRequest(request: ContactRequest) {
        // In a real implementation, this would send the request over the network
        delay(1000) // Simulate network delay
    }
    
    /**
     * Simulate key rotation (for testing)
     */
    private suspend fun simulateKeyRotation(contact: Contact) {
        // In a real implementation, this would request a new public key from the contact
        delay(100) // Simulate network delay
    }
    
    /**
     * Load contacts from storage
     */
    private fun loadContacts() {
        try {
            val contactsJson = encryptedPrefs.getString(CONTACTS_KEY, null)
            if (contactsJson != null) {
                val contactsList = Contact.fromJsonList(contactsJson)
                contacts.clear()
                contactsList.forEach { contact ->
                    contacts[contact.nickname] = contact
                }
                _contactsFlow.value = contacts.values.toList()
            }
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Save contacts to storage
     */
    private fun saveContacts() {
        try {
            val contactsJson = Contact.toJsonList(contacts.values.toList())
            encryptedPrefs.edit()
                .putString(CONTACTS_KEY, contactsJson)
                .apply()
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Load public keys from storage
     */
    private fun loadPublicKeys() {
        try {
            val keysJson = encryptedPrefs.getString(PUBLIC_KEYS_KEY, null)
            if (keysJson != null) {
                val keysMap = PublicKey.fromJsonMap(keysJson)
                publicKeys.clear()
                publicKeys.putAll(keysMap)
            }
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Save public keys to storage
     */
    private fun savePublicKeys() {
        try {
            val keysJson = PublicKey.toJsonMap(publicKeys)
            encryptedPrefs.edit()
                .putString(PUBLIC_KEYS_KEY, keysJson)
                .apply()
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Load contact requests from storage
     */
    private fun loadContactRequests() {
        try {
            val requestsJson = encryptedPrefs.getString(CONTACT_REQUESTS_KEY, null)
            if (requestsJson != null) {
                val requestsList = ContactRequest.fromJsonList(requestsJson)
                contactRequests.clear()
                requestsList.forEach { request ->
                    contactRequests[request.requestId] = request
                }
                _contactRequestsFlow.value = contactRequests.values.toList()
            }
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Save contact requests to storage
     */
    private fun saveContactRequests() {
        try {
            val requestsJson = ContactRequest.toJsonList(contactRequests.values.toList())
            encryptedPrefs.edit()
                .putString(CONTACT_REQUESTS_KEY, requestsJson)
                .apply()
        } catch (e: Exception) {
            // Handle error
        }
    }
    
    /**
     * Validate nickname
     */
    private fun validateNickname(nickname: String) {
        if (nickname.isBlank()) {
            throw ContactManagerException("Nickname cannot be empty")
        }
        if (nickname.length > 32) {
            throw ContactManagerException("Nickname too long (max 32 characters)")
        }
        if (!nickname.matches(Regex("^[a-zA-Z0-9_-]+$"))) {
            throw ContactManagerException("Nickname contains invalid characters")
        }
    }
    
    /**
     * Validate user hash
     */
    private fun validateUserHash(userHash: ByteArray) {
        if (userHash.size != 16) {
            throw ContactManagerException("Invalid user hash size")
        }
    }
    
    /**
     * Generate random bytes
     */
    private fun generateRandomBytes(size: Int): ByteArray {
        val bytes = ByteArray(size)
        Random.nextBytes(bytes)
        return bytes
    }
    
    /**
     * Generate request ID
     */
    private fun generateRequestId(): String {
        return UUID.randomUUID().toString()
    }
    
    /**
     * Cleanup resources
     */
    fun cleanup() {
        keyRotationJob?.cancel()
        contactScope.cancel()
    }
}

// Data Classes

data class Contact(
    val nickname: String,
    val userHash: ByteArray,
    val publicKey: PublicKey,
    val isOnline: Boolean,
    val addedAt: Long,
    val lastSeen: Long,
    val keyRotationCount: Int,
    val lastKeyRotation: Long = 0
) {
    val userHashHex: String
        get() = userHash.joinToString("") { "%02x".format(it) }
    
    companion object {
        fun fromJsonList(json: String): List<Contact> {
            // Implementation for JSON deserialization
            return emptyList()
        }
        
        fun toJsonList(contacts: List<Contact>): String {
            // Implementation for JSON serialization
            return "[]"
        }
    }
}

data class ContactRequest(
    val requestId: String,
    val targetUserHash: ByteArray,
    val requesterNickname: String,
    val requesterPublicKey: PublicKey,
    val timestamp: Long,
    val status: ContactRequestStatus
) {
    val targetUserHashHex: String
        get() = targetUserHash.joinToString("") { "%02x".format(it) }
    
    companion object {
        fun fromJsonList(json: String): List<ContactRequest> {
            // Implementation for JSON deserialization
            return emptyList()
        }
        
        fun toJsonList(requests: List<ContactRequest>): String {
            // Implementation for JSON serialization
            return "[]"
        }
    }
}

data class EncryptedMessage(
    val encryptedData: ByteArray,
    val encryptedKey: ByteArray,
    val iv: ByteArray,
    val recipientNickname: String,
    val timestamp: Long
)

enum class ContactRequestStatus {
    PENDING, ACCEPTED, REJECTED
}

// Extension functions for PublicKey serialization
fun PublicKey.toJsonMap(keys: Map<String, PublicKey>): String {
    // Implementation for JSON serialization
    return "{}"
}

fun PublicKey.fromJsonMap(json: String): Map<String, PublicKey> {
    // Implementation for JSON deserialization
    return emptyMap()
}

class ContactManagerException(message: String, cause: Throwable? = null) : Exception(message, cause)
